akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,293 @@
+          log.debug("[terminator] Initializing termination of server, deadline: {}", PrettyDuration.format(deadline))
+          installTerminationHandlers(deadline.fromNow)
+
+          scheduleOnce(TerminationDeadlineTimerKey, deadline)
+        }
+
+        // this way we expose the termination signal invocation to the external world, in a type safe fashion
+        triggerTermination.success { d ⇒
+          terminateSignal.invoke(d)
+          terminationOfConnectionDone.future // will be completed once termination has completed (in postStop)
+        }
+      }
+
+      setHandler(fromUser, new InHandler {
+        override def onPush(): Unit = {
+          val response = grab(fromUser)
+          pendingUserHandlerResponse = false
+          push(toNet, response)
+        }
+      })
+      setHandler(toUser, new OutHandler {
+        override def onPull(): Unit = {
+          pull(fromNet)
+        }
+      })
+      setHandler(fromNet, new InHandler {
+        override def onPush(): Unit = {
+          val request = grab(fromNet)
+
+          pendingUserHandlerResponse = true
+          push(toUser, request)
+        }
+      })
+      setHandler(toNet, new OutHandler {
+        override def onPull(): Unit = {
+          pull(fromUser)
+        }
+      })
+
+      // TODO what other semantics should we change in this phase...
+      def installTerminationHandlers(deadline: Deadline): Unit = {
+        // we may need to inject a termination response
+        setHandler(fromUser, new InHandler {
+          override def onPush(): Unit = {
+            val overdue = deadline.isOverdue()
+            val response =
+              if (overdue) {
+                log.warning("Terminating server ({}), discarding user reply since arrived after deadline expiration", formatTimeLeft(deadline))
+                settings.terminationDeadlineExceededResponse
+              } else grab(fromUser)
+
+            pendingUserHandlerResponse = false
+            push(toNet, response)
+          }
+        })
+
+        // once termination deadline hits, we stop pulling from network
+        setHandler(toUser, new OutHandler {
+          override def onPull(): Unit = {
+            // if (deadline.hasTimeLeft()) // we pull always as we want to reply errors to everyone
+            pull(fromNet)
+          }
+        })
+
+        setHandler(fromNet, new InHandler {
+          override def onPush(): Unit = {
+            val request = grab(fromNet)
+            log.warning(
+              "Terminating server ({}), attempting to send termination reply to incoming [{} {}]",
+              formatTimeLeft(deadline), request.method, request.uri.path)
+
+            // on purpose discard all incoming bytes for requests
+            // could discard with the deadline.timeLeft completion timeout, but not necessarily needed
+            request.entity.discardBytes()(interpreter.subFusingMaterializer)
```

Comment:
I'm not entirely sure about this... I think it's fine but have not written a test for it yet...
I mean, this is effectively exactly what user code would do, also out-of-bounds in terms of when the next request is pulled for... so should be fine I think... Please sanity check my thinking here @raboof

Reply:
Yes I agree this looks like it makes sense to me.